logsrvd/sendlog.c

```diff
@@ -169,7 +173,66 @@ init_tls_client_context(const char *ca_bundle_file, const char *cert_file, const
 exit:
     return ctx;
 }
-#endif
+
+static bool
+do_tls_connect(struct client_closure *closure)
+{
+    int con_stat, err = -1;
+
+    debug_decl(do_tls_connect, SUDO_DEBUG_UTIL)
+
+    if (ca_bundle == NULL) {
+        sudo_warnx("%s", U_("CA bundle file was not specified"));
+        goto bad;
+    }
+    if (tls_checkpeer && (cert == NULL)) {
+        sudo_warnx("%s", U_("Client certificate was not specified"));
+        goto bad;
+    }
+    if ((ssl_ctx = init_tls_client_context(ca_bundle, cert, key)) == NULL) {
+        sudo_warnx(U_("Unable to initialize ssl context: %s"),
+            ERR_error_string(ERR_get_error(), NULL));
+        goto bad;
+    }
+    if ((ssl = SSL_new(ssl_ctx)) == NULL) {
+        sudo_warnx(U_("Unable to allocate ssl object: %s\n"),
+            ERR_error_string(ERR_get_error(), NULL));
+        goto bad;
+    }
+    if (SSL_set_fd(ssl, closure->sock) <= 0) {
+        sudo_warnx(U_("Unable to attach socket to the ssl object: %s\n"),
+            ERR_error_string(ERR_get_error(), NULL));
+        goto bad;
+    }
+
+    while ((con_stat = SSL_connect(ssl)) != 1) {
```

Comment:
It is not safe to use a static-sized fd_set like this.  If closure->sock is bigger than FD_SETSIZE the FD_SET will overflow fds. poll() is better in this regard or you could adopt the method used by timed_connect() in iolog_client.c

Reply:
i adapted the event based solution